core/eth/state.go

```diff
@@ -1,82 +1,309 @@
+	}
+
+	return cs.getOperatorState(creader, operatorsByQuorum, uint32(blockNumber))
 }
 
-func (cs *ChainState) GetOperatorState(ctx context.Context, blockNumber uint, quorums []core.QuorumID) (*core.OperatorState, error) {
-	operatorsByQuorum, err := cs.Tx.GetOperatorStakesForQuorums(ctx, quorums, uint32(blockNumber))
+// GetCurrentBlockNumber returns the current block number.
+func (cs *ChainState) GetCurrentBlockNumber(creader context.Context) (uint, error) {
+	number, err := cs.Client.BlockNumber(creader)
 	if err != nil {
-		return nil, err
+		cs.logger.Warn("failed to get current block number: %w", "error", err)
+		return 0, fmt.Errorf("failed to get current block number: %w", err)
 	}
 
-	return getOperatorState(operatorsByQuorum, uint32(blockNumber))
+	return uint(number), nil
+}
+
+// GetOperatorSocket returns the socket address for a given operator at the current block number,
+// and it takes blockNumber due to the core.ChainState interface.
+func (cs *ChainState) GetOperatorSocket(creader context.Context, blockNumber uint, operator core.OperatorID) (string, error) {
+	socket, err := cs.Reader.GetOperatorSocket(creader, operator)
+	if err != nil {
+		cs.logger.Warn("failed to get socket for operator %x at block %d: %w", "operator", operator, "blockNumber", blockNumber, "error", err)
+		return "", fmt.Errorf("failed to get socket for operator %x at block %d: %w", operator, blockNumber, err)
+	}
+	return socket, nil
 }
 
-func (cs *ChainState) GetCurrentBlockNumber(ctx context.Context) (uint, error) {
-	header, err := cs.Client.HeaderByNumber(ctx, nil)
+// indexSocketMap filters event logs from the previously checked block number to the current block,
+// to identify all socket update events in that block range, and update the socket map accordingly
+func (cs *ChainState) indexSocketMap(ctx context.Context) error {
+	logs, err := cs.getSocketUpdateEventLogs(ctx)
 	if err != nil {
-		return 0, err
+		return fmt.Errorf("failed to get logs: %w", err)
+	}
+
+	// logs are in order of block number, so we can just iterate through them
+	prematureBreak := false
+	for _, log := range logs {
+		cs.socketPrevBlockNumber.Store(uint32(log.BlockNumber - 1))
+		transaction, err := cs.getTransaction(ctx, log.TxHash)
+		if err != nil {
+			cs.logger.Warn("failed to check transaction %s: %w", "txHash", log.TxHash.Hex(), "error", err)
+			prematureBreak = true
+			break
+		}
+		operatorID, socket, err := cs.parseOperatorSocketUpdate(transaction.Data(), &log)
+		if err != nil {
+			cs.logger.Warn("failed to get transaction data for operator %x: %w", "operatorID", operatorID, "error", err)
+			continue
+		}
+
+		cs.socketMu.Lock()
+		cs.SocketMap[operatorID] = &socket
+		cs.socketMu.Unlock()
+	}
+
+	// If above loop completed without a premature break, increment prevBlockNum by 1 to ensure we don't handle the same block twice
+	if !prematureBreak {
+		socketPrevBlockNumber := cs.socketPrevBlockNumber.Load()
```

Comment:
Instead of loading and then storing, you should use atomic add, `cs.socketPrevBlockNumber.Add(1)`, to prevent a race condition